File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int len=nums.length;
        int round=-1;
        for(int i=len-1;i>=1;i--){
            if(nums[i-1]<nums[i]){
                round=i-1;
                break;
            }
        }
        if(round!=-1){
        for(int j=len-1;j>round;j--){
            if(nums[j]>nums[round]){
                swap(nums,j,round);
                break;
            }
        }}
        reverse(round+1,len-1,nums);
        
    }
    public void swap(int[] nums,int round,int j){
        int temp=nums[j];
        nums[j]=nums[round];
        nums[round]=temp;
    }
    public void reverse(int start,int end,int[] nums){
        while(start<end){
            int temp=nums[start];
            nums[start]=nums[end];
            nums[end]=temp;
            start++;
            end--;    }

    }
}

File: 0031-next-permutation/0031-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        idx=-1
        for i in range(len(nums)-1,0,-1):
            if nums[i-1]<nums[i]:
                idx=i-1
                break
        revIdx=-1
        if idx!=-1:
            for i in range(len(nums)-1,-1,-1):
                if nums[i]>nums[idx]:
                    revIdx=i
                    break
            self.swap(revIdx,idx,nums)
        self.reverse(nums,idx+1,len(nums)-1)
    def swap(self,revIdx,idx,nums):
        temp=nums[revIdx]
        nums[revIdx]=nums[idx]
        nums[idx]=temp
    def reverse(self,nums,st,de):
        while st<de:
            temp=nums[st]
            nums[st]=nums[de]
            nums[de]=temp
            st+=1
            de-=1


        